subsys/pmci/mctp/mctp_i2c_gpio_target.c

```diff
@@ -0,0 +1,225 @@
+	int ret = 0;
+
+	switch (b->reg_addr) {
+	case MCTP_I2C_GPIO_TX_MSG_LEN_ADDR:
+		b->rxtx = true;
+		if (b->tx_pkt == NULL) {
+			LOG_WRN("empty packet?");
+			pkt_len = 0;
+		} else {
+			pkt_len = (uint8_t)(b->tx_pkt->end - b->tx_pkt->start);
+		}
+		*val = pkt_len;
+		break;
+	case MCTP_I2C_GPIO_TX_MSG_ADDR:
+		b->rxtx = true;
+		*val = b->tx_pkt->data[b->tx_pkt->start];
+		b->tx_idx = b->tx_pkt->start;
+		break;
+	default:
+		LOG_WRN("invalid rre reg %d", b->reg_addr);
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+int mctp_i2c_gpio_target_read_processed(struct i2c_target_config *config, uint8_t *val)
+{
+	struct mctp_binding_i2c_gpio_target *b =
+		CONTAINER_OF(config, struct mctp_binding_i2c_gpio_target, i2c_target_cfg);
+
+	b->tx_idx += 1;
+
+	if (b->reg_addr != MCTP_I2C_GPIO_TX_MSG_ADDR) {
+		goto out;
+	}
+
+	if (b->tx_idx > b->tx_pkt->end) {
+		LOG_WRN("rrp past end reg %d", b->reg_addr);
+		return -EIO;
+	}
+
+	*val = b->tx_pkt->data[b->tx_idx];
+
+out:
+	return 0;
+}
+
+int mctp_i2c_gpio_target_stop(struct i2c_target_config *config)
+{
+	struct mctp_binding_i2c_gpio_target *b =
+		CONTAINER_OF(config, struct mctp_binding_i2c_gpio_target, i2c_target_cfg);
+	uint8_t pkt_len;
+
+	if (b->rxtx) {
+		switch (b->reg_addr) {
+		case MCTP_I2C_GPIO_TX_MSG_ADDR:
+			pkt_len = (b->tx_pkt->end - b->tx_pkt->start);
+			if (b->tx_idx < pkt_len - 1) {
+				LOG_WRN("Only %d of %d bytes of the transmit packet were read",
+					b->tx_idx, pkt_len);
+			}
+			b->tx_pkt = NULL;
+			k_sem_give(b->tx_complete);
+			break;
+
+		case MCTP_I2C_GPIO_RX_MSG_ADDR:
+			LOG_DBG("stop rx msg, give pkt");
+			/* Give message to mctp to process */
+			mctp_bus_rx(&b->binding, b->rx_pkt);
+			b->rx_pkt = NULL;
+			break;
+		case MCTP_I2C_GPIO_RX_MSG_LEN_ADDR:
+		case MCTP_I2C_GPIO_TX_MSG_LEN_ADDR:
+			break;
+		default:
+			LOG_WRN("unexpected stop for reg %d", b->reg_addr);
+			break;
+		}
+	}
+
+	return 0;
+}
+
+const struct i2c_target_callbacks mctp_i2c_gpio_target_callbacks = {
+	.write_requested = mctp_i2c_gpio_target_write_requested,
+	.read_requested = mctp_i2c_gpio_target_read_requested,
+	.write_received = mctp_i2c_gpio_target_write_received,
+	.read_processed = mctp_i2c_gpio_target_read_processed,
+	.stop = mctp_i2c_gpio_target_stop,
+};
+
+/*
+ * libmctp wants us to return once the packet is sent not before
+ * so the entire process of flagging the tx with gpio, waiting on the read,
+ * needs to complete before we can move on.
+ *
+ * this is called for each packet in the packet queue libmctp provides
+ */
+int mctp_i2c_gpio_target_tx(struct mctp_binding *binding, struct mctp_pktbuf *pkt)
+{
+	struct mctp_binding_i2c_gpio_target *b =
+		CONTAINER_OF(binding, struct mctp_binding_i2c_gpio_target, binding);
+	int rc;
+
+	k_sem_take(b->tx_lock, K_FOREVER);
+
+	b->tx_pkt = pkt;
+
+	rc = gpio_pin_set_dt(&b->endpoint_gpio, 1);
+	if (rc != 0) {
+		LOG_ERR("failed to set gpio pin");
+		b->tx_pkt = NULL;
+		goto out;
+	}
+
+	k_sem_take(b->tx_complete, K_FOREVER);
```

Comment:
I wonder if we should add a timeout here in case the I2C bus controller does not ever initiate a transfer (e.g: crashes or ??). Would this be something the MCTP endpoint like to know or react against?